miniapps/elasticity/elasticity_driver.cpp

```diff
@@ -0,0 +1,177 @@
+// with stress σ is solved for displacement u.
+//
+//             +----------+----------+
+//   fixed --->|                     |<--- constant displacement
+//             |                     |
+//             +----------+----------+
+//
+// This miniapp uses an elasticity operator that allows for a custom material.
+// By default the NeoHookeanMaterial is used. A linear elastic material is also
+// provided. Based on these examples, other materials could be implemented.
+//
+// The implementation of NeoHookeanMaterial also demenstrates the use of
+// automatic differentiation using either a native dual number forward mode
+// implementation or leveraging the Enzyme third party library.
+
+#include <mfem.hpp>
+
+#include "materials/linear_elastic.hpp"
+#include "materials/neohookean.hpp"
+#include "operators/elasticity_gradient_operator.hpp"
+#include "operators/elasticity_operator.hpp"
+#include "preconditioners/diagonal_preconditioner.hpp"
+
+using namespace std;
+using namespace mfem;
+
+/// This example only works in 3D. Kernels for 2D are not implemented.
+constexpr int dimension = 3;
+
+int main(int argc, char *argv[])
+{
+   Mpi::Init(argc, argv);
+
+   int order = 1;
+   const char *device_config = "cpu";
+   int diagpc_type = ElasticityDiagonalPreconditioner::Type::Diagonal;
+   int serial_refinement_levels = 0;
+   bool paraview = false;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&order, "-o", "--order",
+                  "Finite element order (polynomial degree).");
+   args.AddOption(&device_config, "-d", "--device",
+                  "Device configuration string, see Device::Configure().");
+   args.AddOption(&diagpc_type, "-pc", "--pctype",
+                  "Select diagonal preconditioner type"
+                  " (0:Diagonal, 1:BlockDiagonal).");
+   args.AddOption(&serial_refinement_levels, "-rs", "--ref-serial",
+                  "Number of uniform refinements on the serial mesh.");
+   args.AddOption(&paraview, "-pv", "--paraview", "-no-pv",
+                  "--no-paraview",
+                  "Enable or disable ParaView DataCollection output.");
+   args.ParseCheck();
+
+   Device device(device_config);
+   if (Mpi::Root())
+   {
+      device.Print();
+   }
+
+   auto mesh =
+      Mesh::MakeCartesian3D(8, 2, 2, Element::HEXAHEDRON, 8.0, 1.0, 1.0);
+   if (mesh.Dimension() != dimension)
+   {
+      MFEM_ABORT("This example only works in 3D.");
+   }
+   mesh.EnsureNodes();
+
+   for (int l = 0; l < serial_refinement_levels; l++)
+   {
+      mesh.UniformRefinement();
+   }
+
+   ParMesh pmesh(MPI_COMM_WORLD, mesh);
+   mesh.Clear();
+
+   // Create the elasticity operator on the parallel mesh.
+   ElasticityOperator elasticity_op(pmesh, order);
+
+   // Create and set the material type. We define its GradientType during
+   // instantiation.
+
+   // As seen in materials/gradient_type.hpp there is a choice of the
+   // GradientType with either
+   // * Symbolic (Manually derived)
+   // * EnzymeFwd
```

Comment:
While the name is descriptive, may it wouldn't hurt to say it explicitly, e.g.

```suggestion
   // * EnzymeFwd -- compute gradients using Enzyme's forward mode AD
```